src/Remapping/distributed_remapping.jl

```diff
@@ -144,65 +219,482 @@ function Remapper(target_hcoords, target_zcoords, space)
+        :Remapper,
+    )
+    return Remapper(space, target_hcoords, target_zcoords)
+end
 
 """
-   interpolate(remapper, field; physical_z = false)
+    _set_interpolated_values!(remapper, field)
 
-Interpolate the given `field` as prescribed by `remapper`.
+Change the local state of `remapper` by performing interpolation of `Fields` on the vertical
+and horizontal points.
+"""
+function _set_interpolated_values!(remapper::Remapper, field::Fields.Field)
+    _set_interpolated_values!(
+        remapper._local_interpolated_values,
+        field,
+        remapper._field_values,
+        remapper.local_horiz_indices,
+        remapper.local_horiz_interpolation_weights,
+        remapper.vert_interpolation_weights,
+        remapper.vert_bounding_indices,
+    )
+end
+
+function _set_interpolated_values!(
+    out::AbstractArray,
+    field::Fields.Field,
+    scratch_field_values,
+    local_horiz_indices,
+    (I1, I2)::NTuple{2},
+    vert_interpolation_weights::AbstractArray,
+    vert_bounding_indices::AbstractArray,
+)
+
+    space = axes(field)
+    FT = Spaces.undertype(space)
+    quad = Spaces.quadrature_style(space)
+    Nq = Quadratures.degrees_of_freedom(quad)
+    fv = Fields.field_values(field)
+
+    if ClimaComms.device(field) isa ClimaComms.CUDADevice
+        nblocks, _ = size(I1)
+        nthreads = length(vert_interpolation_weights)
+        @cuda threads = (nthreads) blocks = (nblocks) set_interpolated_values_kernel!(
+            out,
+            (I1, I2),
+            local_horiz_indices,
+            vert_interpolation_weights,
+            vert_bounding_indices,
+            fv,
+        )
+    else
+        # Reading values from fv is expensive, so we try to limit the number of reads. We can do
+        # this because multiple target points might be all contained in the same element.
+        prev_v_lo, prev_v_hi, prev_lidx = -1, -1, -1
+        @inbounds for (vindex, (A, B)) in enumerate(vert_interpolation_weights)
+            (v_lo, v_hi) = vert_bounding_indices[vindex]
+
+            for (out_index, h) in enumerate(local_horiz_indices)
+                # If we are no longer in the same element, read the field values again
+                if prev_lidx != h || v_lo != prev_v_lo || v_hi != prev_v_hi
+                    for j in 1:Nq, i in 1:Nq
+                        scratch_field_values[i, j] = (
+                            A * fv[i, j, nothing, v_lo, h] +
+                            B * fv[i, j, nothing, v_hi, h]
+                        )
+                    end
+                    prev_v_lo, prev_v_hi, prev_lidx = v_lo, v_hi, h
+                end
+
+                tmp = zero(FT)
+
+                for j in 1:Nq, i in 1:Nq
+                    tmp +=
+                        I1[out_index, i] *
+                        I2[out_index, j] *
+                        scratch_field_values[i, j]
+                end
+                out[out_index, vindex] = tmp
+            end
+        end
```

Comment:
Can we put this inside a separate cpu-implementation function?

Reply:
Moved it.